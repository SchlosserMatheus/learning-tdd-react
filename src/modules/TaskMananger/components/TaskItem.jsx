import React, {Component} from 'react';

import PropTypes from 'prop-types';

class TaskItem extends Component
{
    render() {

        const { checked, changeChecked, text} = this.props;
        const textStyle = (checked) ? {color: 'red'} : {color: 'black'};

        return (
            <div>
                <input type="checkbox" onChange={changeChecked} checked={checked}/>
                <span style={textStyle}>{text}</span>
            </div>
        );
    }
}

TaskItem.propTypes = {
    text: PropTypes.string.isRequired,
    checked: PropTypes.bool.isRequired,
    changeChecked: PropTypes.func.isRequired
};

export default TaskItem;