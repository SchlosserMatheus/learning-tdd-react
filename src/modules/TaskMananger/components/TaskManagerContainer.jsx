import React, {Component} from 'react';
import TaskList from "./TaskList";
import TaskItem from "./TaskItem";

class TaskManagerContainer extends Component {

    state = {
        tasks: [
            { text: 'Realizar compras' , checked: false },
            { text: 'Assistir vÃ­deo aulas' , checked: false },
            { text: 'Passar o recado para o joÃ£o' , checked: false },
        ]
    };

    handleChangeChecked = (key, checkedState) => {
        const state = {...this.state};
        state.tasks[key].checked = checkedState;
        const first = [...state.tasks].filter(item => item.checked === false);
        const last = [...state.tasks].filter(item => item.checked === true);
        state.tasks = [...first, ...last];
        this.setState(state);
    };

    render() {
        return (
            <TaskList>
                <h1>TESTE</h1>
                {this.state.tasks.map((item, key) => (
                    <TaskItem key={key}
                      text={item.text}
                      checked={item.checked}
                      changeChecked={event => this.handleChangeChecked(key, !item.checked)}
                    />
                ))}
            </TaskList>
        );
    }
}

export default TaskManagerContainer;