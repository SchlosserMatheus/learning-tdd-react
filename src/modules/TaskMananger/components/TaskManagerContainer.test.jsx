import React from 'react';
import { shallow, mount } from 'enzyme';
import TaskManagerContainer from "./TaskManagerContainer";
import TaskItem from "./TaskItem";

describe('Test TaskManagerContainer Component',  () => {

    it('Check if initial state has three tasks', () => {
        const wrapper = shallow(<TaskManagerContainer/>);
        expect(wrapper.instance().state.tasks.length).toEqual(3);
    });

    /**
     * todo realizar tradução do comentário sobre o reverse dos items quando atualizados com checked true/false
     */
    it('Check if a prop "checked" of a task is updated by method "handleChangeChecked"', () => {
        const wrapper = shallow(<TaskManagerContainer/>);
        const wrapperInstance = wrapper.instance();
        /**
         * os itens que são marcados como checked são jogados para para o final, portanto como todos
         * estão marcados como false, e apenas o item 0 foi marcado como true, ele será jogado para o ultimo item.
         * Então a indice dele de 2 ( porque o estado inicial é criado com 3 items )
         */
        wrapperInstance.handleChangeChecked(0, true);
        expect(wrapperInstance.state.tasks[2].checked).toEqual(true);
    });


    it('dever renderizar tres componentes',() => {
        const wrapper = shallow(<TaskManagerContainer/>);
        expect(wrapper.find('TaskItem')).toHaveLength(3);
    });

    it('Check snapshot version', () => {
        const wrapper = shallow(<TaskManagerContainer/>);
        expect(wrapper).toMatchSnapshot();
    });


});