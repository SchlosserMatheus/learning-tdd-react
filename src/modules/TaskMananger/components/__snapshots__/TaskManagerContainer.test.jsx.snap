// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TaskManagerContainer Component Check snapshot version 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskManagerContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <h1>
          TESTE
        </h1>,
        Array [
          <TaskItem
            changeChecked={[Function]}
            checked={false}
            text="Realizar compras"
          />,
          <TaskItem
            changeChecked={[Function]}
            checked={false}
            text="Assistir vídeo aulas"
          />,
          <TaskItem
            changeChecked={[Function]}
            checked={false}
            text="Passar o recado para o joão"
          />,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "TESTE",
        },
        "ref": null,
        "rendered": "TESTE",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "class",
        "props": Object {
          "changeChecked": [Function],
          "checked": false,
          "text": "Realizar compras",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "changeChecked": [Function],
          "checked": false,
          "text": "Assistir vídeo aulas",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "changeChecked": [Function],
          "checked": false,
          "text": "Passar o recado para o joão",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <h1>
            TESTE
          </h1>,
          Array [
            <TaskItem
              changeChecked={[Function]}
              checked={false}
              text="Realizar compras"
            />,
            <TaskItem
              changeChecked={[Function]}
              checked={false}
              text="Assistir vídeo aulas"
            />,
            <TaskItem
              changeChecked={[Function]}
              checked={false}
              text="Passar o recado para o joão"
            />,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "TESTE",
          },
          "ref": null,
          "rendered": "TESTE",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "changeChecked": [Function],
            "checked": false,
            "text": "Realizar compras",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "changeChecked": [Function],
            "checked": false,
            "text": "Assistir vídeo aulas",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "changeChecked": [Function],
            "checked": false,
            "text": "Passar o recado para o joão",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
